backend/server.py

```
@@ -17,16 +35,351 @@
+                    
+                    if tool_name != last_tool_used:  # Only notify about new tools
+                        last_tool_used = tool_name
+                        tool_data = {"tool_use": {"tool_name": tool_name}}
+                        yield f"data: {json.dumps(tool_data)}\n\n"
+                else:
+                    # Regular text chunks
+                    response_data = {"chunk": chunk} if isinstance(chunk, str) else chunk
+                    yield f"data: {json.dumps(response_data)}\n\n"
+        except Exception as e:
+            error_msg = str(e)
+            tb = traceback.format_exc()
```

Ruff (0.8.2)

Undefined name `traceback`

(F821)